File: scripts/ui/dialogs/JsonDialog.js
const Editor = require('ui/dialogs/Editor')

let dialog, label, p
exports.load = function () {
	dialog = new Dialog
	label = new Label('')
	p = new Table
	p.center();
	p.defaults().padTop(10).left();
	p.add('$editor.sourceCode', Color.gray).padRight(10).padTop(0).row();
	p.table(cons(t => {
		t.right();
		t.button(Icon.download, Styles.clearPartiali, run(() => Vars.ui
			.showConfirm(
				'粘贴', '是否要粘贴', run(() => {
					this.file.writeString(Core.app.getClipboardText());
					label.setText(this.getText());
				}))
		));
		t.button(Icon.copy, Styles.clearPartiali, run(() => {
			Core.app.setClipboardText(this.file.readString());
		}));
	})).growX().right().row();
	p.pane(cons(p => p.left().add(label))).width(bw).height(bh);
	dialog.cont.add(p).grow().row();

	dialog.buttons.button('$back', Icon.left, Styles.defaultt, run(() => {
		dialog.hide();
	})).size(bw / 2, 55);

	let listener = extend(VisibilityListener, {
		hidden: () => {
			this.file = Editor.file
			dialog.title.setText(this.file.nameWithoutExtension())
			label.setText(this.getText())
			Editor.ui.removeListener(listener)
			return false;
		}
	})
	dialog.buttons.button('$edit', Icon.edit, Styles.defaultt, run(() => {
		Editor.edit(this.file, this.mod)
		Editor.ui.addListener(listener)
	})).size(bw / 2, 55);
	dialog.closeOnBack();
}
exports.getText = function () {
	return this.file.readString().replace(/\r/g, '\n').replace(/\t/g, '  ').replace(/\[\s*([^]*?)\s*\]/g, '[ $1 ]')
}

let w = Core.graphics.getWidth(),
	h = Core.graphics.getHeight(),
	bw = w > h ? 550 : 450,
	bh = w > h ? 200 : Vars.mobile ? 300 : 350;
exports.constructor = function (file, mod) {
	if (!/^h?json$/.test(file.extension())) return null;
	this.file = file, this.mod = mod

	dialog.title.setText(file.name() != null ? file.name() : '');

	label.setText(this.getText());

	dialog.show();
	return dialog;
}

File: src/modmake/ui/dialog/JsonDialog.java
package modmake.ui.dialog;

import arc.Core;
import arc.files.Fi;
import arc.graphics.Color;
import arc.scene.event.VisibilityListener;
import arc.scene.ui.Label;
import arc.scene.ui.layout.Table;
import mindustry.Vars;
import mindustry.gen.Icon;
import mindustry.ui.Styles;
import modmake.components.*;
import modmake.ui.MyStyles;
import modmake.util.tools.FilterTable;

public class JsonDialog extends Window {
	Label label;
	Table p;
	Fi    file;
	MyMod mod;
	float w = Core.graphics.getWidth(),
	 h      = Core.graphics.getHeight(),
	 bw     = w > h ? 550 : 450,
	 bh     = w > h ? 200 : Vars.mobile ? 300 : 350;

	public JsonDialog() {
		super("json", 120, 80, true, false);
	}


	public void load() {
		label = new Label("");
		p = new Table();
		p.update(() -> {
			if (file == null || !file.exists()) hide();
		});
		p.center();
		p.defaults().padTop(10).left();
		p.add("@editor.sourceCode", Color.gray).padRight(10).padTop(0).row();
		p.table(t -> {
			t.right();
			t.button(Icon.paste, MyStyles.clearPartiali, () -> Vars.ui
			 .showConfirm("@paste", "@confirm.paste", () -> {
				 file.writeString(Core.app.getClipboardText());
				 label.setText(getText());
			 })
			).padRight(2);
			t.button(Icon.copy, MyStyles.clearPartiali, () -> {
				Core.app.setClipboardText(this.file.readString());
			});
		}).growX().right().row();
		p.pane(p -> p.left().add(label).grow()).minSize(bw, bh).grow();
		cont.add(p).grow().row();

		// buttons.button("@back", Icon.left, Styles.defaultt, () -> {
		// 	hide();
		// }).size(bw / 2, 55);

		Editor[] editor = {null};
		hidden(() -> {
			if (editor[0] != null) editor[0].find(t -> {
				if (t instanceof FilterTable) {
					Core.app.post(t::clear);
				}
				return true;
			});
		});
		var listener = new VisibilityListener() {
			public boolean hidden() {
				file = editor[0].file;
				title.setText(file.nameWithoutExtension());
				editor[0].file = null;
				label.setText(getText());
				editor[0].removeListener(this);
				return false;
			}
		};
		buttons.button("@edit", Icon.edit, Styles.defaultt, () -> {
			 checkset(editor);
			 editor[0].edit(file, mod);
			 editor[0].addListener(listener);
		 }).size(bw / 2f, 55)
		 .disabled(b -> editor[0] != null && editor[0].isShown());
		// buttons.row();
		buttons.button("文本编辑", Icon.edit, Styles.defaultt, () -> {
			 checkset(editor);
			 editor[0].edit(file, mod, true);
			 editor[0].addListener(listener);
		 }).size(bw / 2f, 55)
		 .disabled(b -> editor[0] != null && editor[0].isShown());

		// closeOnBack();
	}
	private static void checkset(Editor[] editor) {
		if (editor[0] == null) {
			editor[0] = new Editor();
			editor[0].load();
		}
	}
	public String getText() {
		return file.readString().replaceAll("\\r", "\n").replaceAll("\\t", "  ")
		 .replaceAll("\\[(#?\\w*)]", "[\u0001$1]");
	}

	public JsonDialog show(Fi file, MyMod mod) {
		if (!file.extEquals("hjson") && !file.extEquals("json")) return null;
		this.file = file;
		this.mod = mod;

		title.setText(file.name() != null ? file.name() : "");

		label.setText(getText());

		show();
		return this;
	}
}
